### PR TITLE
Eigenvalue and eigenvector power iteration method

### PR DESCRIPTION
**Pull request by antonAce**
**Power iteration method** - eigenvalue numeric algorithm, based on recurrent relation: L(i+1) = (A * l(i)) / || A * l(i) ||, where l(i) - eigenvector approximation. Also I added package [Fluent Assertions](https://github.com/fluentassertions/fluentassertions) to the project and used in algorithm tests. Related issue #111.
 
References:
- N. N. Kalitkin "Numeric methods" p.186

- [x] I have performed a self-review of my code
- [x] My code follows the style guidelines of this project
- [x] I have added comments to hard-to-understand areas of my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have made corresponding changes to the comments
- [x] I have made corresponding changes to the README.md

### CODE REVIEW

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Utilities/Extensions/VectorExtensions.cs  3
- Algorithms.Tests/LinearAlgebra/Eigenvalue/PowerIterationTests.cs  1
- Algorithms/LinearAlgebra/Eigenvalue/PowerIteration.cs  4
- Utilities/Extensions/RandomExtensions.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6319749&bid=20543568)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- Utilities/Extensions/RandomExtensions.cs  1
- Algorithms/LinearAlgebra/Eigenvalue/PowerIteration.cs  4
- Utilities/Extensions/VectorExtensions.cs  3
- Algorithms.Tests/LinearAlgebra/Eigenvalue/PowerIterationTests.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6319749&bid=20543568)

File: README.md
Comment (siriak) [inline README.md:24]: Codacy found an issue: [[list-item-bullet-indent] Incorrect indentation before bullet: remove 2 spaces](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6319749)

Comment (antonAce) [reply README.md]: Fixed!

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- Utilities/Extensions/VectorExtensions.cs  3
- Algorithms/LinearAlgebra/Eigenvalue/PowerIteration.cs  4
- Algorithms.Tests/LinearAlgebra/Eigenvalue/PowerIterationTests.cs  1
- Utilities/Extensions/RandomExtensions.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6319749&bid=20543568)

File: README.md
Comment (siriak) [inline README.md:25]: Codacy found an issue: [[list-item-bullet-indent] Incorrect indentation before bullet: remove 4 spaces](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6319749)

Comment (antonAce) [reply README.md:25]: @siriak, Codacy lint is always throwing an issue about **Incorrect indentation**, even when tab format is correct. Is this a bug?

Comment (asiriak) [reply README.md:25]: Try selecting the line with a mouse and you'll see that there are spaces at the beginning of the lines and tabs are used everywhere else

Comment (antonAce) [reply README.md]: Oh, I see. Fixed!

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Utilities/Extensions/VectorExtensions.cs  3
- Algorithms/LinearAlgebra/Eigenvalue/PowerIteration.cs  4
- Algorithms.Tests/LinearAlgebra/Eigenvalue/PowerIterationTests.cs  1
- Utilities/Extensions/RandomExtensions.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6319749&bid=20543568)

File: Algorithms/LinearAlgebra/Eigenvalue/PowerIteration.cs
Comment (siriak) [inline Algorithms/LinearAlgebra/Eigenvalue/PowerIteration.cs:21]:
```suggestion
        /// <description>The <paramref name="source"/> matrix must be square-shaped.</description>

```

File: Utilities/Extensions/RandomExtensions.cs
Comment (siriak) [inline Utilities/Extensions/RandomExtensions.cs:15]: There are some extensions/utilities for vectors and matrices in #174, so we might want to wait a bit until it's merged and see what can be reused from or added there. So that we don't create a lot of methods that all do the same thing.

File: Utilities/Extensions/VectorExtensions.cs
Comment (siriak) [inline Utilities/Extensions/VectorExtensions.cs:5]: And same goes here

**Review by siriak — COMMENTED** [review]
Looks good, waiting on #174 to be merged to avoid duplication

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Utilities/Extensions/RandomExtensions.cs  1
- Utilities/Extensions/VectorExtensions.cs  3
- Algorithms.Tests/LinearAlgebra/Eigenvalue/PowerIterationTests.cs  1
- Algorithms/LinearAlgebra/Eigenvalue/PowerIteration.cs  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6319749&bid=20543568)

File: Utilities/Extensions/RandomExtensions.cs
Comment (antonAce) [reply Utilities/Extensions/RandomExtensions.cs:15]: Yep, in #174 there are some duplicated methods, so when it's merged, I'll rebase my changes and reuse existing logic. But I insist on replacing static methods with extension methods for vector/matrix operations and more some methods from SVD to Utilities (like **RandomUnitVector**).

Comment (siriak) [reply Utilities/Extensions/RandomExtensions.cs:15]: #174 is merged

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Utilities/Extensions/RandomExtensions.cs  1
- Algorithms.Tests/LinearAlgebra/Eigenvalue/PowerIterationTests.cs  1
- Algorithms/LinearAlgebra/Eigenvalue/PowerIteration.cs  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6319749&bid=20543568)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Utilities/Extensions/RandomExtensions.cs  1
- Algorithms/LinearAlgebra/Eigenvalue/PowerIteration.cs  4
- Utilities.Tests/Extensions/MatrixExtensionsTests.cs  1
- Algorithms.Tests/LinearAlgebra/Eigenvalue/PowerIterationTests.cs  1
         

Complexity decreasing per file
==============================
+ Utilities/Extensions/MatrixExtensions.cs  -2
         

Clones added
============
- Utilities.Tests/Extensions/MatrixExtensionsTests.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6319749&bid=20543568)

File: Utilities/Extensions/RandomExtensions.cs
Comment (antonAce) [reply Utilities/Extensions/RandomExtensions.cs]: @siriak reused existing logic and covered extensions with unit tests

File: Utilities/Extensions/VectorExtensions.cs
Comment (siriak) [inline Utilities/Extensions/VectorExtensions.cs:23]: Not an extension

File: Utilities/Extensions/VectorExtensions.cs
Comment (siriak) [inline Utilities/Extensions/VectorExtensions.cs:17]: Do you think this method is necessary? IMO it can be removed, only 1 line

File: Utilities/Extensions/VectorExtensions.cs
Comment (siriak) [inline Utilities/Extensions/VectorExtensions.cs:40]: Same

File: Utilities/Extensions/VectorExtensions.cs
Comment (siriak) [inline Utilities/Extensions/VectorExtensions.cs:61]: Same

File: Utilities/Extensions/VectorExtensions.cs
Comment (siriak) [inline Utilities/Extensions/VectorExtensions.cs:82]: Same

File: Utilities/Extensions/VectorExtensions.cs
Comment (siriak) [inline Utilities/Extensions/VectorExtensions.cs:95]: Same

File: Algorithms.Tests/Numeric/Decomposition/SVDTests.cs
Comment (siriak) [inline Algorithms.Tests/Numeric/Decomposition/SVDTests.cs:1]: Have you changed CRLF to LF or why is this changed?

**Review by siriak — COMMENTED** [review]
Looks good, I've left some comments about extension methods

File: Utilities/Extensions/VectorExtensions.cs
Comment (antonAce) [reply Utilities/Extensions/VectorExtensions.cs:17]: It's used in the SVD algorithm

File: Algorithms.Tests/Numeric/Decomposition/SVDTests.cs
Comment (antonAce) [reply Algorithms.Tests/Numeric/Decomposition/SVDTests.cs:1]: Yes, this is an IDE formatting

File: Utilities/Extensions/VectorExtensions.cs
Comment (antonAce) [reply Utilities/Extensions/VectorExtensions.cs]: Replaced with extension

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Algorithms/LinearAlgebra/Eigenvalue/PowerIteration.cs  4
- Algorithms.Tests/LinearAlgebra/Eigenvalue/PowerIterationTests.cs  1
- Utilities/Extensions/RandomExtensions.cs  1
- Utilities.Tests/Extensions/MatrixExtensionsTests.cs  1
         

Complexity decreasing per file
==============================
+ Utilities/Extensions/MatrixExtensions.cs  -2
         

Clones added
============
- Utilities.Tests/Extensions/MatrixExtensionsTests.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6319749&bid=20543568)

File: Utilities/Extensions/VectorExtensions.cs
Comment (antonAce) [reply Utilities/Extensions/VectorExtensions.cs]: Replaced with extension

File: Utilities/Extensions/VectorExtensions.cs
Comment (antonAce) [reply Utilities/Extensions/VectorExtensions.cs]: Replaced with extension

File: Utilities/Extensions/VectorExtensions.cs
Comment (antonAce) [reply Utilities/Extensions/VectorExtensions.cs]: Replaced with extension

File: Utilities/Extensions/VectorExtensions.cs
Comment (antonAce) [reply Utilities/Extensions/VectorExtensions.cs]: Replaced with extension

**Review by siriak — APPROVED** [review]
Looks good, thanks!

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Utilities.Tests/Extensions/MatrixExtensionsTests.cs  1
- Algorithms/LinearAlgebra/Eigenvalue/PowerIteration.cs  4
- Utilities/Extensions/RandomExtensions.cs  1
- Algorithms.Tests/LinearAlgebra/Eigenvalue/PowerIterationTests.cs  1
         

Complexity decreasing per file
==============================
+ Utilities/Extensions/MatrixExtensions.cs  -2
         

Clones added
============
- Utilities.Tests/Extensions/MatrixExtensionsTests.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6319749&bid=20543568)

**Pull request merged**
